Align https-by-default upgrade logic with mixed-content upgrading

sysrqb opened the pull request:
#### 678ece22c6542a1ae99472c11d2204bd93be7ea1
<pre>
Align https-by-default upgrade logic with mixed-content upgrading
<a href="https://bugs.webkit.org/show_bug.cgi?id=277053">https://bugs.webkit.org/show_bug.cgi?id=277053</a>
<a href="https://rdar.apple.com/132444853">rdar://132444853</a>

Reviewed by Alex Christensen.

This patch performs https upgrade for an initial resource request, and on
redirect, similar to mixed content. This patch also leverages
ResourceRequest::wasSchemeOptimisticallyUpgraded() to make a decision about
whether we are in a redirect loop, instead of re-evaluating whether
https-by-default is enabled.

I also moved setting a shorter time-out for optimistically upgraded requests
into the NetworkResourceLoader. In the future I want to experiment with setting
a value that is more appropriate for the current network, and tracking those
measurements should happen in the network process.

Also I&apos;m disabling the preference in WTR. Some tests are failing on some ports,
but not all. I want to keep this as testable.

No new tests, covered by existing tests, and:
    1) move from [webView _mainFrameURL] to [webView URL]
    2) Adjust one subtest to account for moving the upgrading behavior for
    redirects into CachedResourceLoader::updateRequestAfterRedirection.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::changeLocation):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::shouldPerformHTTPSUpgrade):
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::shouldUpgradeRequestforHTTPSOnly const): Deleted.
(WebCore::CachedResourceLoader::upgradeRequestforHTTPSOnlyIfNeeded const): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, HTTPSFirstHTTPDowngrade)):
(TEST(WKNavigation, HTTPSFirstHTTPDowngradeRedirect)):
(TEST(WKNavigation, HTTPSFirstLocalHostIPAddress)):
(TEST(WKNavigation, HTTPSOnlyInitialLoad)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallback)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackBypassEnabledCertificateError)):
(TEST(WKNavigation, HTTPSOnlyWithSameSiteBypass)):
(TEST(WKNavigation, HTTPSOnlyWithHTTPRedirect)):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/281429@main">https://commits.webkit.org/281429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/317ae917cb444512328e20fb8080c884b5babbb0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12409 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63796 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10403 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62008 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10580 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7269 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61909 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/36584 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51845 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29389 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33290 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9084 "Found 1 new test failure: imported/w3c/web-platform-tests/shadow-dom/focus/focus-pseudo-matches-on-shadow-host.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9326 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9366 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65526 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3807 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55889 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51827 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56032 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3162 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8968 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35042 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -761,28 +761,20 @@ FetchMetadataSite CachedResourceLoader::computeFetchMetadataSiteAfterRedirection
-        return true;
-    }
-    return false;
+        && newURL.host().contains("."_s)
```
This seems strange.  IPv4 and domains should have a dot, but IPv6 doesn't.  Is that really what you want?

sysrqb replied:
This is for single-label host names. Those generally won't support https. We have a separate exception for IP addresses and localhost in `ResourceRequestBase::upgradeInsecureRequestIfNeeded`. I debated adding this check there, as well, but then it interferes with upgrade-insecure-request logic - but that might be okay, too.

sysrqb replied:
I decided to just remove this. It was a small addition, but not necessary for this PR.

achristensen07 commented on Source/WebCore/loader/ResourceLoader.cpp:
```diff
@@ -449,8 +449,7 @@ void ResourceLoader::willSendRequestInternal(ResourceRequest&& request, const Re
     if (frameLoader() && frameLoader()->frame().isMainFrame() && cachedResource() && cachedResource()->type() == CachedResource::Type::MainResource && !redirectResponse.isNull()) {
-        auto requestURL { redirectResponse.url() };
-        if (documentLoader() && documentLoader()->cachedResourceLoader().upgradeRequestforHTTPSOnlyIfNeeded(requestURL, request) && request.url() == redirectResponse.url()) {
+        if (request.wasSchemeOptimisticallyUpgraded() && request.url() == redirectResponse.url()) {
```
This is commenting on code that isn't changing, but is this to detect redirect loops?  This won't detect redirect loops where two URLs are redirecting back and forth.  Usually we'll instead use redirectCount

sysrqb replied:
The original implementation had two purposes. 1) it performed the scheme upgraded, if needed, and 2) it short-circuited the redirect count and detected when the request URL is cycling directly between http and https for the same resource.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1950,16 +1941,15 @@ HTTPServer httpServer({
-    EXPECT_EQ(loadCount, 2);
-    EXPECT_WK_STREQ(@"http://site2.example/secure3", [webView _mainFrameURL].absoluteString);
+    EXPECT_EQ(loadCount, 3);
+    EXPECT_WK_STREQ(@"http://site2.example/secure2", [webView _mainFrameURL].absoluteString);
```
Could you explain what caused this change in behavior?
Also, why use _mainFrameURL instead of WKWebView.URL?

sysrqb replied:
I don't remember why I used _mainFrameURL, probably lack of knowledge. I can change that.

The change is behavior is a consequence the upgrading happening in CachedResourceLoader::updateRequestAfterRedirection instead in ResourceLoader::willSendRequestInternal when we were detecting the redirect loop. I can make add a note about this in the commit message

Commit:
Align https-by-default upgrade logic with mixed-content upgrading
https://bugs.webkit.org/show_bug.cgi?id=277053
rdar://132444853

Reviewed by Alex Christensen.

This patch performs https upgrade for an initial resource request, and on
redirect, similar to mixed content. This patch also leverages
ResourceRequest::wasSchemeOptimisticallyUpgraded() to make a decision about
whether we are in a redirect loop, instead of re-evaluating whether
https-by-default is enabled.

I also moved setting a shorter time-out for optimistically upgraded requests
into the NetworkResourceLoader. In the future I want to experiment with setting
a value that is more appropriate for the current network, and tracking those
measurements should happen in the network process.

Also I'm disabling the preference in WTR. Some tests are failing on some ports,
but not all. I want to keep this as testable.

No new tests, covered by existing tests, and:
    1) move from [webView _mainFrameURL] to [webView URL]
    2) Adjust one subtest to account for moving the upgrading behavior for
    redirects into CachedResourceLoader::updateRequestAfterRedirection.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::changeLocation):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::shouldPerformHTTPSUpgrade):
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::shouldUpgradeRequestforHTTPSOnly const): Deleted.
(WebCore::CachedResourceLoader::upgradeRequestforHTTPSOnlyIfNeeded const): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, HTTPSFirstHTTPDowngrade)):
(TEST(WKNavigation, HTTPSFirstHTTPDowngradeRedirect)):
(TEST(WKNavigation, HTTPSFirstLocalHostIPAddress)):
(TEST(WKNavigation, HTTPSOnlyInitialLoad)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallback)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackBypassEnabledCertificateError)):
(TEST(WKNavigation, HTTPSOnlyWithSameSiteBypass)):
(TEST(WKNavigation, HTTPSOnlyWithHTTPRedirect)):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: https://commits.webkit.org/281429@main
